app/state.go

```diff
@@ -1104,6 +1109,68 @@ func (fir *fetchStoreRetryState) Handle(ctx *StateContext, c Controller) (State,
 	return fir.Wait(NewUpdateFetchState(&fir.update), fir, intvl, ctx.WakeupChan)
 }
 
+type inventoryUpdateRetry struct {
+	baseState
+	WaitState
+	from State
+	err  error
+}
+
+func NewInventoryUpdateRetryState(from State, update *datastore.UpdateInfo,
+	err error) State {
+	return &inventoryUpdateRetry{
+		baseState: baseState{
+			id: datastore.MenderStateInventoryUpdateRetryWait,
+			t:  ToSync,
+		},
+		WaitState: NewWaitState(datastore.MenderStateInventoryUpdateRetryWait, ToSync),
+		from:      from,
+		err:       err,
+	}
+}
+
+func (fir *inventoryUpdateRetry) Cancel() bool {
+	return fir.WaitState.Cancel()
+}
+func (fir *inventoryUpdateRetry) Handle(ctx *StateContext, c Controller) (State, bool) {
+	if ctx.nextAttemptAt.After(time.Now()) {
+		secondsRemaining := time.Duration(ctx.nextAttemptAt.Unix()-time.Now().Unix()) * time.Second
+		if secondsRemaining.Seconds() > ctx.inventoryPollInterval.Seconds() {
+			secondsRemaining = ctx.inventoryPollInterval
+		}
+		log.Infof("Handle update inventory retry state: pass, not the time yet; %ds remaining.",
+			secondsRemaining/time.Second)
+		return fir.Wait(NewInventoryUpdateState(), fir, secondsRemaining, ctx.WakeupChan)
+	}
+	log.Infof("Handle update inventory retry state try: %d", ctx.inventoryUpdateAttempts)
+	err := c.InventoryRefresh()
+	if err != nil {
+		log.Warnf("Failed to refresh inventory: %v", err)
+		if errors.Cause(err) == errNoArtifactName {
+			return NewErrorState(NewTransientError(err)), false
+		}
+	}
+
+	interval, err := client.GetExponentialBackoffTime(
+		ctx.inventoryUpdateAttempts,
+		c.GetUpdatePollInterval(),
+		ctx.retryCount,
+	)
+	ctx.nextAttemptAt = time.Now().Add(interval)
+	if err != nil {
+		log.Infof("Handle update inventory retry state: failed to send inventory: %s", err.Error())
+		return States.CheckWait, false
+	}
+
+	ctx.inventoryUpdateAttempts++
+
+	if interval.Seconds() > ctx.inventoryPollInterval.Seconds() {
+		interval = ctx.inventoryPollInterval
+	}
+	log.Infof("Wait %v before next inventory update attempt", interval)
+	return fir.Wait(NewInventoryUpdateState(), fir, interval, ctx.WakeupChan)
```

Comment:
As far as I can see, there is no success condition here. It will always wait, regardless of whether the refresh worked or not.

Reply:
right. that was really a preliminary call for a review. thanks!